File: app/src/main/java/it/chalmers/gamma/adapter/primary/web/ThrottlingController.java
package it.chalmers.gamma.adapter.primary.web;

import it.chalmers.gamma.app.throttling.ThrottlingManagementFacade;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.servlet.ModelAndView;

@Controller
public class ThrottlingController {

  private final ThrottlingManagementFacade throttlingManagementFacade;

  public ThrottlingController(ThrottlingManagementFacade throttlingManagementFacade) {
    this.throttlingManagementFacade = throttlingManagementFacade;
  }

  @GetMapping("/throttling")
  public ModelAndView getThrottling(
      @RequestHeader(value = "HX-Request", required = false) boolean htmxRequest) {
    ModelAndView mv = new ModelAndView();
    if (htmxRequest) {
      mv.setViewName("throttling/throttling");
    } else {
      mv.setViewName("index");
      mv.addObject("page", "throttling/throttling");
    }

    mv.addObject("throttling", throttlingManagementFacade.getAll());

    return mv;
  }

  @DeleteMapping("/throttling/{key}")
  public ModelAndView deleteActivationCode(
      @RequestHeader(value = "HX-Request", required = false) boolean htmxRequest,
      @PathVariable("key") String key) {
    this.throttlingManagementFacade.deleteKey(key);

    return new ModelAndView("common/empty");
  }
}

File: app/src/main/kotlin/it/chalmers/gamma/ThrottlingController.kt
@file:Suppress("ForbiddenVoid") // Spring uses Void for redirects with no response body.

package it.chalmers.gamma

import it.chalmers.gamma.platform.html.csrfInput
import it.chalmers.gamma.platform.html.gammaPage
import it.chalmers.gamma.platform.html.methodOverrideInput
import it.chalmers.gamma.throttling.ThrottleKey
import it.chalmers.gamma.throttling.ThrottlingAdministration
import jakarta.servlet.http.HttpServletRequest
import kotlinx.html.FormMethod
import kotlinx.html.button
import kotlinx.html.form
import kotlinx.html.table
import kotlinx.html.tbody
import kotlinx.html.td
import kotlinx.html.tr
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.security.web.csrf.CsrfToken
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController

@RestController
class ThrottlingController(
    private val throttling: ThrottlingAdministration,
) {
    @GetMapping("/throttling", produces = [MediaType.TEXT_HTML_VALUE])
    fun throttling(
        authentication: Authentication,
        csrfToken: CsrfToken,
        request: HttpServletRequest,
    ): String {
        val page = pageContext(authentication, csrfToken, request)
        val snapshot = throttling.snapshot(authentication.actor())
        return gammaPage("Throttling", page) {
            table {
                tbody {
                    snapshot.entries.forEach { entry ->
                        tr {
                            td { +entry.key.value }
                            td { +entry.attempts.toString() }
                            td {
                                form(
                                    action = "${page.contextPath}/throttling/${entry.key.value}",
                                    method = FormMethod.post,
                                ) {
                                    csrfInput(csrfToken.token)
                                    methodOverrideInput("delete")
                                    button { +"Delete" }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    @DeleteMapping("/throttling/{key}")
    fun delete(
        authentication: Authentication,
        @PathVariable key: String,
    ): ResponseEntity<Void> {
        throttling.delete(authentication.actor(), ThrottleKey(key))
        return redirect("/throttling")
    }
}
